src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+  ((parents :type list
+            :initarg :class-parents
+            :reader class-parents)
+   (attributes :type list
+               :initarg :class-attributes
+               :reader class-attributes)))
+
+(defclass variable-reference (reference)
+  ((value :initarg :variable-reference-value
+          :reader variable-reference-value)))
+
+
+(defclass misc-reference (reference)
+  ((custom-type :initarg :misc-custom-type
+                :type string
+                :reader misc-custom-type)))
+
+(defstruct capture-regex regex function)
+
+(defclass detective-search () ())
+
+(defclass search-regex (detective-search)
+  ((function-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :function-regex
+                   :writer set-function-regex
+                   :reader search-function-regex)
+   (package-regex :type (or capture-regex null)
+                  :initform nil
+                  :initarg :package-regex
+                  :writer set-package-regex
+                  :reader search-package-regex)
+   (class-regex :type (or capture-regex null)
+                :initform nil
+                :initarg :class-regex
+                :writer set-class-regex
+                :reader search-class-regex)
+   (variable-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :variable-regex
+                   :writer set-variable-regex
+                   :reader search-variable-regex )
+   (misc-regex :type (or capture-regex null)
+               :initform nil
+               :initarg :misc-regex
+               :writer set-misc-regex
+               :reader search-misc-regex)))
+
+
+(defun buffer-references (buffer)
+  (buffer-value buffer 'references))
+
+(defun (setf buffer-references) (value buffer)
+  (setf (buffer-value buffer 'references) value))
+
+(defgeneric search-references (search-class))
+
+(defmethod search-references ((search search-regex))
+  (labels ((find-ref (regex class)
+             (with-point ((p (buffer-start-point (point-buffer (current-point)))))
+               (loop :for position = (search-forward-regexp p (capture-regex-regex regex))
+                     :while position
+                     :collect (funcall (capture-regex-function regex) 
+                                       (copy-point position)
+                                       class)))))
+  (with-slots (function-regex
+               package-regex
+               class-regex
+               variable-regex
+               misc-regex)
+      search
+    (let ((slots 
+            (list (cons (cons "functions" function-regex):function-reference)
+                  (cons (cons "classes" class-regex) :class-reference)
+                  (cons (cons "packages" package-regex) :package-reference)
+                  (cons (cons "variables" variable-regex) :variable-reference)
+                  (cons (cons "misc" misc-regex) :misc-reference))))
```

Comment:
If it is an ID, wouldn't a keyword be preferable to a string?
This may be a matter of taste.

Reply:
I tend to use strings for no particular reason, if you want I can change it to a keyword, I have no problem.